grails-app/migrations/0.9.x/changelog-2024-08-20-0000-migrate-prepayment-invoices-add-amount-and-inverse-items.groovy

```diff
@@ -0,0 +1,126 @@
+ *
+ * More info on the structure of this file: https://grails.github.io/grails-database-migration/latest/index.html#_general_format
+ */
+databaseChangeLog = {
+
+    changeSet(author: "ewaterman", id: "200820240000-0") {
+
+        //noinspection GroovyAssignabilityCheck
+        grailsChange {
+            //noinspection GroovyAssignabilityCheck
+            change {
+
+                PrepaymentInvoiceService prepaymentInvoiceService = ctx.getBean("prepaymentInvoiceService")
+
+                /*
+                 * Set the inverse field for all invoice items that existed before the partial invoicing
+                 * feature was introduced.
+                 *
+                 * Any inverse invoice items that are defined in the database will have the inverse field also defined
+                 * (before the field was introduced, inverse items were not stored in the database at all and instead
+                 * were calculated dynamically when fetched). As such, it's safe to assume that if inverse is not
+                 * defined for an invoice item, it's not an inverse invoice item.
+                 */
+                InvoiceItem.findAllByInverseIsNull().each { invoiceItem ->
+                    invoiceItem.inverse = false
+                    invoiceItem.save(failOnError: true, flush: true)
+                }
+
+                InvoiceType prepaymentInvoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
+
+                /*
+                 * Set the amount field for all prepayment invoice items that existed before the partial invoicing
+                 * feature was introduced. (The amount field will be null for all pre-existing invoices since the field
+                 * was not being used.)
+                 */
+                Invoice.findAllByInvoiceType(prepaymentInvoiceType).each { prepaymentInvoice ->
+                    prepaymentInvoice.invoiceItems.each { invoiceItem ->
+                        // TODO: This should probably be put in a helper method in PrepaymentInvoiceService
+                        BigDecimal prepaymentPercent = (invoiceItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
+                        invoiceItem.amount = invoiceItem.quantity * invoiceItem.unitPrice * prepaymentPercent
+                        invoiceItem.save(failOnError: true, flush: true)
+                    }
+                }
+
+                /*
+                 * For every pre-existing prepayment invoice, for every invoice item, add an inverse invoice item
+                 * to the final invoice of the order.
+                 *
+                 * For existing data from before the partial invoicing feature was introduced, ALL non-prepayment
+                 * invoices will ALWAYS be the final invoice, meaning every order will have one prepayment invoice
+                 * and one non-prepayment (aka final) invoices.
```

Comment:
I'm second guessing myself. Is this even true? Against develop I was just able to generate two separate invoices for an order with two items, one invoice per item...

Reply:
:thinking: Did you try it on the staging too? I think there might be some initial implementation of this feature on the develop (allowing to generate it, so perhaps that why you could do it)